### ISSUE TITLE
Thread-unsafe access to string interner in BinaryWire

### ISSUE THREAD

**Issue by mcs6502** [OpenHFT/Chronicle-Wire#137]
BinaryWire uses its string interners (Bit8 and UTF8) in a thread-unsafe manner. This causes the interners to return inconsistent results when multiple threads simultaneously perform text operations involving the BinaryWire class.

**Comment by dpisklov** [comment]
Wire (any implementation) is not thread safe as a class, so it should never be used in multiple threads. This is intentional (mainly for performance reasons).

**Comment by mcs6502** [comment]
Please see attached patch [BinaryWireStringInternerTest.patch.txt](https://github.com/OpenHFT/Chronicle-Wire/files/3003555/BinaryWireStringInternerTest.patch.txt) for a unit test reproducing this issue. (It is a combination of shouldInternExistingStrings() and multipleThreadsUsingBinaryWiresShouldNotCauseProblems() from the same class.)

**Comment by mcs6502** [comment]
Hi Dmitry, you might want to review this issue a bit further -- the two existing tests mentioned in the previous comment don't share any wires and should be able to run simultaneously (going by the name of the second test). But thanks for looking anyways!
~~~~
java.lang.AssertionError: At index 5 for string puccinspmmiguwjvxbbeuilhuipgjbibctymlmaxpxbwrlmugcqvcsylxnfplvvonyakkklshdpnicipohlrloiyrbjwqmmqyifyllecopvfbbawxewofskjyvgbgpwuugmddxbxneovuggsgawjpavqijthlxixxduprodughssjcyfynrnufwpybmxxtmtnhrnagugvchgklboflglpujgwyvtcrpaedymicqryyklt
Expected: is <76636201>
     but: was <677514035>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at net.openhft.chronicle.wire.BinaryWireStringInternerTest.shouldInternExistingStringsAlright(BinaryWireStringInternerTest.java:79)
~~~~

**Comment by dpisklov** [comment]
You are right actually, apologies. I reopened the issue

**Comment by peter-lawrey** [comment]
The API for Wire requires external locking if you want thread safety.
As many calls are made to read or write a single object, making individual methods thread safe is more deceptive than useful.

**Comment by peter-lawrey** [comment]
StringInterner only guarantees it will behave in a correct manner.
It doesn't guarantee that all threads see the same data and that multiple threads will return the same String object for the same string. It is designed to be a best-effort basis so it can be as lightweight as possible.

**Comment by mcs6502** [comment]
ok, thanks for reviewing this guys, and for clarifying the available guarantees in StringInterner.

**Comment by dpisklov** [comment]
@mcs6502 thanks for raising this!

**Comment by hft-team-city** [comment]
Released in [Chronicle-Wire-2.20.101](https://github.com/OpenHFT/Chronicle-Wire/releases/chronicle-wire-2.20.101), [BOM-2.20.134](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.20.134)
